src/levanter/data/shardcache.py

```diff
@@ -0,0 +1,712 @@
+# Dataset for preprocessing data, tokenizing, and caching to disk.
+import asyncio
+import logging
+import sys
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from typing import IO, Dict, Generic, Iterator, List, Optional, Protocol, Sequence, Tuple, TypeVar
+
+import fsspec.core
+import pyarrow as pa
+import pyarrow.parquet as pq
+import ray
+import tblib
+from dataclasses_json import dataclass_json
+from tqdm import tqdm
+
+from levanter.data import Dataset
+
+
+# We want to support the following:
+#  1) Deterministic ordering over the data, even for a changing number of readers or writers
+#  2) Sharded reading and writing
+#  3) Simultaneous reading and writing of shards
+#  4) Fast resumption without losing too much progress
+#  5) (eventually) shuffling/random access
+
+# What does consumption look like?
+# There are two or three use cases:
+# 1) We have a larger training dataset, and we want to draw samples from it more or less independently on a large number
+# of machines. We don't really care about "epochs"/"passes", but we do want to be able to handle resumes and be
+# deterministic
+# 2) We have a smaller validation dataset, and we want to do a single pass over it. We don't care about resuming.
+# 3) Like (1) but we want to jump around the dataset. We still care about resuming and determinism, but don't
+# care about epochs.
+#
+#
+# We focus on (1) and (2) for now.
+
+# What does our input look like?
+# We have a smallish number of input shards (though usually >= num machines), each of which is an iterator over documents.
```

Comment:
Does this assume the shards are uniform i.e. that we have a single Hugging Face dataset, or Parquet / JSONL file glob? Would this support concatenating multiple datasets?

Reply:
atm i'm assuming either a single HF dataset, or a list of jsonl globs

Reply:
for multiple datasets, I think the right thing to do is cache them separately and do mixing on the fly, since that's a thing one might want to play around with